### PR TITLE
Fix misleading builtin command syntax of /shutdown

### PR DESCRIPTION
**Pull request by Wuzzy2**
I noticed the `/shutdown` command syntax help was incorrect and misleading. This is the syntax help:

> [\<delay_in_seconds> | -1] [reconnect] [\<message>]

But this was misleading:

* The `[reconnect]` argument actually expected a literal `true` to enable the reconnect (even if the syntax help would be correct, that's still unintuitive)
* The `[reconnect]` argument also was not actually optional; you could not go to the message without specifying this argument

I fixed this:

* `[reconnect]` now actually expects the literal word “reconnect” and is also optional
* If the word “reconnect” is specified, reconnect is activated.
* If the word is not specified, any text after the `delay_in_seconds` will be read as the message text
* No strings are touched

## How to test

* Add a dummy debug log in the builtin code to test the parsed parameter values before `core.request_shutdown` is called.
* Call `/shutdown` in-game an play around with different parameters:
    * /shutdown 5 reconnect Hello World
    * /shutdown 5 Hello World
    * /shutdown 5

### CODE REVIEW

**Comment by SmallJoker** [comment]
A literal argument is not user-friendly, even less now that the command arguments might be translated. I'd suggest to do the opposite: adjust the command description to the behaviour of the function so that server admins can use `y/yes/1/true` to specify the "reconnect" arg.

**Comment by Wuzzy2** [comment]
I think the current “real” syntax with “true”/etc. is pretty confusing as well. How do you explain that well in the syntax? The syntax could literally just include the word “true” which doesn't explain anything at all.

I think this shows a limitation in the current chat command systems: There is not enough room to explain every single argument. :-/

Another idea would be to throw away the current bad syntax and come up with a better one from scratch. Suggestions welcome.

**Comment by Wuzzy2** [comment]
How about this syntax?

```
[<delay_in_seconds> | -1] [-r] [<message>]
```

And specify what “-r” means in the help text.

**Comment by Wuzzy2** [comment]
Okay, I just applied the syntax I suggested above.

File: builtin/game/chat.lua [hunk 1]
```diff
@@ -983,23 +983,31 @@ core.register_chatcommand("days", {
...
-		message = param and param:match("^%s*(.+)") or ""
+		local bool_reconnect = false
+		if reconnect == "-r" then
+			bool_reconnect = true
+		else
+			message = reconnect
+		end
+		if not message then
+			message = param and param:match("^%s*(.+)") or ""
+		end
```
Comment (SmallJoker) [inline builtin/game/chat.lua:1003]: When `delay` or `reconnect` are not provided (or invalid), the text is currently cut off to the first word.
I suggest better pattern matching, using a separate variable in case matching failed to to omitted arguments.

**Comment by Wuzzy2** [comment]
@SmallJoker: Good testing. Yes, the parameter parsing was indeed still broken, whoops. I have changed the parsing completey to allow for any possible combination of arguments (hopefully), in compliance with the syntax help.

I tested with a "dry run" method by commenting out `core.request_shutdown` and adding this test code at the end (to see which args will actually get fed into `core.request_shutdown`):
```
core.log("error", "delay="..tostring(delay).." reconnect="..tostring(bool_reconnect).." message:trim()="..message:trim())
```

**Comment by SmallJoker** [comment]
Thinking about it, I might have found an even simpler approach:

```
local fields = param:split(" ")

local timeout = -1
if tonumber(fields[1]) then
	-- Seconds value is provided
	timeout = tonumber(fields[1])
	table.remove(fields, 1)
end

local reconnect = false
if fields[1] == "-r" then
	reconnect = true
	table.remove(fields, 1)
end

local message = table.concat(fields, " ")
```

This however strips double spaces in the message, if that's acceptable.

**Comment by Wuzzy2** [comment]
> This however strips double spaces in the message, if that's acceptable.

No, that's not acceptable. Don't mess around with the user input for no good reason.

**Comment by SmallJoker** [comment]
> No, that's not acceptable. 

Okay, in this case it would be `local fields = param:split(" ", true)` where the last argument also includes empty strings. The input will not be altered and the duplicated code + manual string concat is then also superfluous.

**Comment by Wuzzy2** [comment]
@SmallJoker: Your proposed change still doesn't work for the following command, even with the new split arg:

```
/shutdown 55      -r message bla text toe om
```

(e.g. multiple spaces before `-r`)

**Review by SmallJoker — APPROVED** [review]
Works

File: builtin/game/chat.lua [hunk 2]
```diff
@@ -983,23 +983,60 @@ core.register_chatcommand("days", {
...
+				one = param:match("(.*)")
+				if tonumber(one) then
+					delay = one
+				elseif one == "-r" then
+					reconnect = one
+				else
+					message = one
+				end
+			end
+		end
```
Comment (sfan5) [inline builtin/game/chat.lua:1025]: Please move this to a function so early return can be used. The logic is quite hard to follow here.

**Comment by Wuzzy2** [comment]
@sfan5: Done.

File: builtin/game/chat.lua [hunk 3]
```diff
@@ -982,24 +982,66 @@ core.register_chatcommand("days", {
...
+			if tonumber(one) then
+				delay = one
+			elseif one == "-r" then
+				reconnect = one
+			else
+				message = one
+			end
+		end
+	end
+	return delay, reconnect, message
```
Comment (sfan5) [inline builtin/game/chat.lua:1022]: You moved it but the logic is just as hard to follow as before.
By "early return" I mean replacing line 991 with `return one, two, three`, the rest of function can then be taken out of the else clause (same thing in other applicable places)

File: builtin/game/chat.lua [hunk 4]
```diff
@@ -982,24 +982,66 @@ core.register_chatcommand("days", {
...
 	func = function(name, param)
-		local delay, reconnect, message
-		delay, param = param:match("^%s*(%S+)(.*)")
-		if param then
-			reconnect, param = param:match("^%s*(%S+)(.*)")
+		local delay, reconnect, message = parse_shutdown_param(param)
+		local bool_reconnect = false
+		if reconnect == "-r" then
+			bool_reconnect = true
+		end
```
Comment (sfan5) [inline builtin/game/chat.lua:1034]:
```suggestion
		local bool_reconnect = reconnect == "-r"
```

**Comment by Wuzzy2** [comment]
Okay, @sfan5, I think I understood now, and made the updates (and tested). Please check again.
